python/lsst/ip/diffim/registerImage.py

```diff
@@ -0,0 +1,192 @@
+        @param inputWcs: Wcs of input exposure
+        @param inputBBox: Bounding box of input exposure
+        @param templateSources: Sources from template exposure
+        @return Struct(matches: Matches between sources,
+                       wcs: Wcs for input in frame of template,
+                       )
+        """
+        matches = self.matchSources(inputSources, templateSources)
+        wcs = self.fitWcs(matches, inputWcs, inputBBox)
+        return Struct(matches=matches, wcs=wcs)
+
+    def matchSources(self, inputSources, templateSources):
+        """Match sources between the input and template
+
+        The order of the input arguments matters (because the later Wcs
+        fitting assumes a particular order).
+
+        @param inputSources: Source catalog of the input frame
+        @param templateSources: Source of the target frame
+        @return Match list
+        """
+        matches = afwTable.matchRaDec(templateSources, inputSources,
+                                      self.config.matchRadius*afwGeom.arcseconds)
+        self.log.info("Matching within %.1f arcsec: %d matches" % (self.config.matchRadius, len(matches)))
+        self.metadata.set("MATCH_NUM", len(matches))
+        if len(matches) == 0:
+            raise RuntimeError("Unable to match source catalogs")
+        return matches
+
+    def fitWcs(self, matches, inputWcs, inputBBox):
+        """Fit Wcs to matches
+
+        The fitting includes iterative sigma-clipping.
+
+        @param matches: List of matches (first is target, second is input)
+        @param inputWcs: Original input Wcs
+        @param inputBBox: Bounding box of input image
+        @return Wcs
+        """
+        copyMatches = type(matches)(matches)
+        refCoordKey = copyMatches[0].first.getTable().getCoordKey()
+        inCentroidKey = copyMatches[0].second.getTable().getCentroidKey()
+        for i in range(self.config.sipIter):
+            sipFit = makeCreateWcsWithSip(copyMatches, inputWcs, self.config.sipOrder, inputBBox)
+            self.log.debug("Registration WCS RMS iteration %d: %f pixels",
+                           i, sipFit.getScatterInPixels())
+            wcs = sipFit.getNewWcs()
+            dr = [m.first.get(refCoordKey).angularSeparation(
+                wcs.pixelToSky(m.second.get(inCentroidKey))).asArcseconds() for
+                m in copyMatches]
+            dr = numpy.array(dr)
+            rms = math.sqrt((dr*dr).mean())  # RMS from zero
+            rms = max(rms, 1.0e-9)  # Don't believe any RMS smaller than this
+            self.log.debug("Registration iteration %d: rms=%f", i, rms)
+            good = numpy.where(dr < self.config.sipRej*rms)[0]
+            numBad = len(copyMatches) - len(good)
+            self.log.debug("Registration iteration %d: rejected %d", i, numBad)
+            if numBad == 0:
+                break
+            copyMatches = type(matches)(copyMatches[i] for i in good)
```

Comment:
This should be much faster: replace `np.where` above with `good = dr < self.config.sipRej*rms`, and the creation of the new `copyMatches` with `copyMatches = list(itertools.compress(x,a))` (I believe type(matches) is a list in this instance, but you can check that: if it's more like a np array you can just slice it directly with the bool array).

You also would replace `len(good)` with `good.sum()`, since it's a bool array now.